ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiWorldStateUpdater.java

```diff
@@ -219,76 +217,82 @@ public void commit() {
-        } else {
-          pendingValue.setUpdated(value);
-        }
-      }
-      updatedAccount.getUpdatedStorage().clear();
-
-      if (pendingStorageUpdates.isEmpty()) {
-        storageToUpdate.remove(updatedAddress);
-      }
-
-      if (tracked.getStorageWasCleared()) {
-        tracked.setStorageWasCleared(false); // storage already cleared for this transaction
-      }
-
-      // TODO maybe add address preimage?
-    }
+    getUpdatedAccounts().parallelStream()
+        .forEach(
+            tracked -> {
+              final Address updatedAddress = tracked.getAddress();
+              final BonsaiAccount updatedAccount;
+              if (tracked.getWrappedAccount() == null) {
+                final BonsaiValue<BonsaiAccount> updatedAccountValue =
+                    accountsToUpdate.get(updatedAddress);
+                updatedAccount = new BonsaiAccount(this, tracked);
+                tracked.setWrappedAccount(updatedAccount);
+                if (updatedAccountValue == null) {
+                  accountsToUpdate.put(updatedAddress, new BonsaiValue<>(null, updatedAccount));
+                  codeToUpdate.put(
+                      updatedAddress, new BonsaiValue<>(null, updatedAccount.getCode()));
+                } else {
+                  updatedAccountValue.setUpdated(updatedAccount);
+                }
+              } else {
+                updatedAccount = tracked.getWrappedAccount();
+                updatedAccount.setBalance(tracked.getBalance());
+                updatedAccount.setNonce(tracked.getNonce());
+                if (tracked.codeWasUpdated()) {
+                  updatedAccount.setCode(tracked.getCode());
+                }
+                if (tracked.getStorageWasCleared()) {
+                  updatedAccount.clearStorage();
+                }
+                tracked.getUpdatedStorage().forEach(updatedAccount::setStorageValue);
+              }
+
+              if (tracked.codeWasUpdated()) {
+                final BonsaiValue<Bytes> pendingCode =
+                    codeToUpdate.computeIfAbsent(
+                        updatedAddress,
+                        addr ->
+                            new BonsaiValue<>(wrappedWorldView().getCode(addr).orElse(null), null));
+                pendingCode.setUpdated(updatedAccount.getCode());
+              }
+
+              final Map<Hash, BonsaiValue<UInt256>> pendingStorageUpdates =
+                  storageToUpdate.computeIfAbsent(updatedAddress, __ -> new ConcurrentHashMap<>());
+              if (tracked.getStorageWasCleared()) {
+                storageToClear.add(updatedAddress);
+                pendingStorageUpdates.clear();
+              }
+
+              final TreeSet<Map.Entry<UInt256, UInt256>> entries =
+                  new TreeSet<>(Map.Entry.comparingByKey());
+              entries.addAll(updatedAccount.getUpdatedStorage().entrySet());
+
+              entries.forEach(
```

Comment:
why not `entries.parallelStream().forEach(` ?  can we not parallelize the storage update also  or do we think there are diminishing returns here?

Reply:
This was one of the tests we did, you can find it if you take a look to the first commit in this PR. We had with the database corrupted with this test.

Reply:
In this case could we maybe add a comment why we are not using `parallelStream`? I think somebody could see this code and try it again. With a comment warning that it will not work, we should save time.